engine/buff/buff.cpp

```diff
@@ -829,6 +831,76 @@ buff_t* buff_t::set_duration_multiplier( double multiplier )
   return this;
 }
 
+buff_t* buff_t::set_time_modifier( double modifier )
+{
+  if ( s_data->flags( spell_attribute::SX_IGNORE_FOR_MOD_TIME_RATE ) )
+    return this;
+
+  modifier = std::max( modifier, -99.0 ); // Limit slow down to 100x slower
+
+  if ( modifier == time_modifier )
+    return this;
+
+  auto old_modifier = time_modifier;
+  time_modifier = modifier;
+
+  if ( current_stack <= 0 || expiration.empty() )
+  {
+    return this;
+  }
+
+  auto ratio = ( old_modifier + 100.0 ) / ( time_modifier + 100.0 );
+
+  // Slowing down the clock, expiry will be moved into the future
+  if ( time_modifier < old_modifier )
+  {
+    for ( auto exp : expiration )
+    {
+      auto new_remains = exp->remains() * ratio;
+      exp->reschedule( new_remains );
+
+      if ( player )
+        sim->print_log( "{} increases {} expiry by a factor of {}. New expiration: {}", *player, *this, ratio,
+                      exp->occurs() );
+    }
+  }
+  // Speeding up the clock, expiry moved closer
+  else
+  {
+    auto old_expirations = new std::vector<event_t*>();
```

Comment:
This leaks memory. Is this vector even necessary? Seems like you could just change the elements of the vector rather than create a new one.